### PR TITLE
Implement Auto Flush

### PR DESCRIPTION
**Pull request by dvonthenen**
## Proposed changes

This finished off the implementation of the Finalize functionality. Issue: https://github.com/deepgram/deepgram-python-sdk/issues/351

By implementing the Python version of the auto flush found in the Go SDK here:
https://github.com/deepgram/deepgram-go-sdk/pull/237

## Types of changes

What types of changes does your code introduce to the community Python SDK?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update or tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I have lint'ed all of my code using repo standards
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

What is interesting to note is the time difference in processing between the Threaded and Async IO Tasks. Async IO Tasks can be on SEVERAL orders of magnitude **SLOWER** than the Threaded client. It was originally the reason why I wanted to deprecate the Async Client because, for IO-intensive apps, it flat out just doesn't make any sense.

### CODE REVIEW

File: deepgram/options.py [hunk 1]
```diff
@@ -82,6 +100,43 @@
...
+
+        bool_auto_flush_delta_new = False
+        auto_flush_delta_new = self.options.get("keep_alive")
+        if isinstance(auto_flush_delta_new, bool):
+            bool_auto_flush_delta_new = auto_flush_delta_new
+        else:
+            bool_auto_flush_delta_new = auto_flush_delta_new == "true"
+
+        return bool_auto_flush_delta_org or bool_auto_flush_delta_new
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/options.py:122]: sweet! thanks coderabbit!

File: tests/edge_cases/auto_flush/async_microphone_mute/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Live API (Real-Time) Example
+
+This example uses the Microphone as input in order to detect conversation insights in what is being said. This example required additional components (for the microphone) to be installed in order for this example to function correctly.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/async_microphone_mute/README.md:3]: since this is a test for an edge case, I am just doing to delete the READMEs here

File: tests/edge_cases/auto_flush/microphone_mute/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Live API (Real-Time) Example
+
+This example uses the Microphone as input in order to detect conversation insights in what is being said. This example required additional components (for the microphone) to be installed in order for this example to function correctly.
+
+## Prerequisites
+
+This example will only work on Linux and MacOS. Windows platforms are not supported.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/microphone_mute/README.md:7]: since this is a test for an edge case, I am just doing to delete the READMEs here

File: tests/edge_cases/auto_flush/microphone_mute/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# Live API (Real-Time) Example
+
+This example uses the Microphone as input in order to detect conversation insights in what is being said. This example required additional components (for the microphone) to be installed in order for this example to function correctly.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/microphone_mute/README.md:3]: since this is a test for an edge case, I am just doing to delete the READMEs here

File: tests/edge_cases/auto_flush/microphone_mute/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+You must add your `DEEPGRAM_API_KEY` to your list of environment variables. We use environment variables because they are easy to configure, support PaaS-style deployments, and work well in containerized environments like Docker and Kubernetes.
+
+```bash
+export DEEPGRAM_API_KEY=YOUR-APP-KEY-HERE
+```
+
+## Installation
+
+The Live API (Real-Time) example makes use of a [microphone package](https://github.com/deepgram/deepgram-python-sdk/tree/main/deepgram/audio/microphone) contained within the repository. That package makes use of the [PortAudio library](http://www.portaudio.com/) which is a cross-platform open source audio library. If you are on Linux, you can install this library using whatever package manager is available (yum, apt, etc.) on your operating system. If you are on macOS, you can install this library using [brew](https://brew.sh/).
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/microphone_mute/README.md:21]: since this is a test for an edge case, I am just doing to delete the READMEs here

File: deepgram/clients/live/v1/client.py [hunk 6]
```diff
@@ -561,6 +715,31 @@
...
+            if not ret:
+                self._logger.error("keep_alive failed")
+                self._logger.spam("LiveClient.keep_alive LEAVE")
+                return False
+
+        self._logger.notice("keep_alive succeeded")
+        self._logger.spam("LiveClient.keep_alive LEAVE")
+
+        return True
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/live/v1/client.py:742]: Yep, there is one path that isn't handled. good catch!

File: tests/edge_cases/auto_flush/async_microphone_mute/README.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+You must add your `DEEPGRAM_API_KEY` to your list of environment variables. We use environment variables because they are easy to configure, support PaaS-style deployments, and work well in containerized environments like Docker and Kubernetes.
+
+```bash
+export DEEPGRAM_API_KEY=YOUR-APP-KEY-HERE
+```
+
+## Installation
+
+The Live API (Real-Time) example makes use of a [microphone package](https://github.com/deepgram/deepgram-python-sdk/tree/main/deepgram/audio/microphone) contained within the repository. That package makes use of the [PortAudio library](http://www.portaudio.com/) which is a cross-platform open source audio library. If you are on Linux, you can install this library using whatever package manager is available (yum, apt, etc.) on your operating system. If you are on macOS, you can install this library using [brew](https://brew.sh/).
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/async_microphone_mute/README.md:21]: since this is a test for an edge case, I am just doing to delete the READMEs here

File: tests/edge_cases/auto_flush/async_microphone_mute/README.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# Live API (Real-Time) Example
+
+This example uses the Microphone as input in order to detect conversation insights in what is being said. This example required additional components (for the microphone) to be installed in order for this example to function correctly.
+
+## Prerequisites
+
+This example will only work on Linux and MacOS. Windows platforms are not supported.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/async_microphone_mute/README.md:7]: since this is a test for an edge case, I am just doing to delete the READMEs here

File: deepgram/options.py [hunk 9]
```diff
@@ -82,6 +100,43 @@
...
+        """
+        is_auto_flush_enabled: Returns True if the client is configured to auto-flush.
+        """
+        auto_flush_delta = self.options.get("auto_flush_reply_delta")
+        return (
+            auto_flush_delta is not None
+            and isinstance(auto_flush_delta, numbers.Number)
+            and auto_flush_delta > 0
+        )
+
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/options.py:133]: I need to set the return type which should be a bool.

File: deepgram/options.py [hunk 10]
```diff
@@ -27,6 +28,16 @@ class DeepgramClientOptions:
...
 
+    _logger: verboselogs.VerboseLogger
+
+    verbose: int
+    url: str
+    api_key: str
+    headers: Dict[str, str]
+    options: Dict[str, str]
+
+    _inspect: bool
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/options.py:39]: removing this declaration. was just testing something out.

File: deepgram/clients/live/v1/client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/live/v1/client.py]: If you look at `self.send()`, we are catching the exceptions within that function and the return type is `bool`. You should have no exception coming out of that function.

File: deepgram/options.py [hunk 11]
```diff
@@ -27,6 +28,9 @@
         options: (Optional) Additional options for initializing the client.
     """
 
+    _logger: verboselogs.VerboseLogger
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/options.py:31]: I was done in __init__, but it does read better on first initialization.

File: tests/edge_cases/auto_flush/microphone_mute/main.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                "keepalive": "true",
+                "auto_flush_reply_delta": 2000,
+            },
+        )
+        deepgram: DeepgramClient = DeepgramClient("", config)
+
+        dg_connection = deepgram.listen.live.v("1")
+
+        def on_open(self, open, **kwargs):
+            print(f"Connection Open")
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply tests/edge_cases/auto_flush/microphone_mute/main.py:39]: I think you mentioned this below...

File: deepgram/clients/live/v1/client.py [hunk 13]
```diff
@@ -6,6 +6,7 @@
 import time
 import logging
 from typing import Dict, Union, Optional, cast, Any
+from datetime import datetime
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/live/v1/client.py:9]: Verified!

File: deepgram/clients/live/v1/client.py [hunk 14]
```diff
@@ -189,18 +216,19 @@ def on(
...
-        self._logger.info("event fired: %s", event)
+        self._logger.info("event subscribed: %s", event)
         if event in LiveTranscriptionEvents.__members__.values() and callable(handler):
             self._event_handlers[event].append(handler)
 
     def _emit(self, event: LiveTranscriptionEvents, *args, **kwargs) -> None:
         """
         Emits events to the registered event handlers.
         """
+        self._logger.debug("callback handlers for: %s", event)
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply deepgram/clients/live/v1/client.py:227]: I prefer having a separate method as it is easier to read

**Review by SandraRodgers — APPROVED** [review]
